### ISSUE TITLE
TSX files doesn't suggest component properties

### ISSUE THREAD

**Issue by nschurmann** [microsoft/TypeScript#29017]
<img width="576" alt="screen shot 2018-10-01 at 1 25 00 pm" src="https://user-images.githubusercontent.com/5099344/46301684-6c199580-c57d-11e8-9c20-05990508c6dc.png">
This are the only props suggested in the autocomplete, @types for react router are installed, so it should suggest at least path and component.

**Comment by DanielRosenwasser** [comment]
Repro:

```tsx
import * as React from "react";
import { Route } from "react-router"

class App extends React.Component {
	render() {
		return (
			<div>
				<Route path="" /*1*/ />
				<Route exact /*2*/ />
				<Route exact={true} /*3*/ />
			</div>
		);
	}
}
```

For the 3 markers, (1) has a completion that includes `""` which is totally nonsensical, and all 3 exhibit the explained behavior.

![image](https://user-images.githubusercontent.com/972891/49992226-4d9a2800-ff38-11e8-8d2f-8a941e391e67.png)

**Comment by weswigham** [comment]
Non-JSX repro of the same issue:
```ts
interface Foo {
    a?: number;
    b?: number;
}

function f<T extends Foo>(x: T) { }

f({/*1*/});
```

branch with a candidate fix (so I don't forget) is [here](https://github.com/weswigham/TypeScript/pull/new/contextual-typing-no-signature-resolution). Has some issues with the `ThisType<T>` tag that I need to investigate before opening a PR and seems to regress services behavior for
```ts
window.addEventListener("error", event/*1*/ => {
	event.filename;
});
```
but not
```ts
interface EventListenerWithOverloads {
	addEventListener(type: "error", listener: (this: Window, ev: ErrorEvent) => any, options?: boolean | AddEventListenerOptions): void;
	addEventListener(type: "beforeunload", listener: (this: Window, ev: BeforeUnloadEvent) => any, options?: boolean | AddEventListenerOptions): void;
}

declare var ev: EventListenerWithOverloads;

ev.addEventListener("error", event/*1*/ => {
	event.filename;
})
```
which I'll need to look into as well.

I've long held that `getContextualType` needed to not be reentrant with inference, and with that branch, that is the case. The effect is that completions are a lot better (they come from the uninstantiated signatures now) and that literal types are retained in calls with errors and that an inferred boolean type can no longer contextual type itself. We've had a bunch of issues that essentially boil down to this same underlying flaw in contextual typing - I'll need to gather them all up along with their test cases.

**Comment by weswigham** [comment]
This has been fixed since #36556 went in.
